File: AlgorithmLcr/Try3-Component types with known topology and corresponding positions/topofit_id/iofmt.py
"""Unified inputs (../../INPUT_FORMAT.md): measurements + known topology.

Measurements: first line n, then n lines ``f Rz Iz`` (sec.1).
Topology: the parameter-free form of the unified adjacency matrix (sec.2.3)
-- node count V, then the upper-triangle rows of edge counts (row i holds
V-1-i integers for slots (i,i+1)..(i,V-1)), then the edge-type queue, one
R|L|C per line in slot-major order.  Reconstruction walks the slots in row
order and consumes queue entries to emit edges (i, j, kind) with i < j.

``#`` starts a comment (whole line or trailing), blank lines are ignored.
Dump uses %.17g so load(dump(x)) == x bit-for-bit.
"""

from __future__ import annotations

import numpy as np


def _content_lines(text: str) -> list[str]:
    """Strip comments and blank lines (INPUT_FORMAT.md sec.0)."""
    out = []
    for raw in text.splitlines():
        s = raw.split("#", 1)[0].strip()
        if s:
            out.append(s)
    return out


# ---------------------------------------------------------------------------
# measurements (sec.1)
# ---------------------------------------------------------------------------

def format_measurements(f, z) -> str:
    """Serialize (f, z) to the unified text form."""
    f = np.asarray(f, dtype=float)
    z = np.asarray(z, dtype=complex)
    if len(f) != len(z):
        raise ValueError(f"f has {len(f)} points, z has {len(z)}")
    if len(f) == 0:
        raise ValueError("need at least one measurement point")
    lines = [str(len(f))]
    for fk, zk in zip(f, z):
        if not (np.isfinite(fk) and fk > 0.0):
            raise ValueError(f"frequency must be positive and finite, got {fk}")
        if not (np.isfinite(zk.real) and np.isfinite(zk.imag)):
            raise ValueError(f"impedance parts must be finite, got {zk}")
        lines.append(f"{fk:.17g} {zk.real:.17g} {zk.imag:.17g}")
    return "\n".join(lines) + "\n"


def parse_measurements(text: str) -> tuple[np.ndarray, np.ndarray]:
    """Parse the unified text form -> (f, z).  Validation per sec.3."""
    rows = _content_lines(text)
    if not rows:
        raise ValueError("empty measurement input")
    try:
        n = int(rows[0])
    except ValueError:
        raise ValueError(f"first line must be integer n, got {rows[0]!r}") from None
    if n < 1:
        raise ValueError(f"n must be a positive integer, got {n}")
    data = rows[1:]
    if len(data) != n:
        raise ValueError(f"n={n} but {len(data)} data lines follow")
    fs, reals, imags = [], [], []
    for k, line in enumerate(data, start=2):  # 1-based line no. incl. n-line
        parts = line.split()
        if len(parts) != 3:
            raise ValueError(f"line {k}: expected 'f Rz Iz', got {line!r}")
        try:
            fv, rz, iz = (float(x) for x in parts)
        except ValueError:
            raise ValueError(f"line {k}: non-numeric field in {line!r}") from None
        if not np.isfinite([fv, rz, iz]).all():
            raise ValueError(f"line {k}: non-finite value in {line!r}")
        if fv <= 0.0:
            raise ValueError(f"line {k}: frequency must be > 0, got {fv}")
        fs.append(fv)
        reals.append(rz)
        imags.append(iz)
    return (np.asarray(fs, dtype=float),
            np.asarray(reals) + 1j * np.asarray(imags))


def load_measurements(path) -> tuple[np.ndarray, np.ndarray]:
    """Read the unified measurement file from disk."""
    with open(path, "r", encoding="utf-8") as fh:
        return parse_measurements(fh.read())


# ---------------------------------------------------------------------------
# topology: adjacency count matrix + edge-type queue (sec.2.3)
# ---------------------------------------------------------------------------

def _slot_pairs(V: int) -> list[tuple[int, int]]:
    """Upper-triangle slots in row-major order: (0,1),(0,2),...,(V-2,V-1)."""
    return [(i, j) for i in range(V) for j in range(i + 1, V)]


def format_topology(edges: list) -> str:
    """Serialize an edge list [(u, v, kind), ...] in the unified form.

    V = max node label + 1.  Edges are stably ordered slot-major (same-slot
    edges keep their input order), then dumped as count rows + type queue.
    """
    if not edges:
        raise ValueError("need at least one edge")
    V = max(max(u, v) for u, v, _k in edges) + 1
    ordered = sorted(edges, key=lambda e: (min(e[0], e[1]), max(e[0], e[1])))
    counts: dict[tuple[int, int], int] = {}
    for u, v, _k in ordered:
        counts[(min(u, v), max(u, v))] = counts.get((min(u, v), max(u, v)), 0) + 1
    lines = [str(V)]
    for i in range(V - 1):
        lines.append(" ".join(str(counts.get((i, j), 0))
                              for j in range(i + 1, V)))
    for _u, _v, k in ordered:
        lines.append(k)
    return "\n".join(lines) + "\n"


def parse_topology(text: str) -> list:
    """Parse the topology file -> edges [(u, v, kind), ...] with u < v.

    Validation per INPUT_FORMAT.md sec.3; connectivity / dead parts are the
    engine's business (reduce_graph), not the loader's.
    """
    rows = _content_lines(text)
    if not rows:
        raise ValueError("empty topology input")
    try:
        V = int(rows[0])
    except ValueError:
        raise ValueError(f"first line must be integer V, got {rows[0]!r}") from None
    if V < 2:
        raise ValueError(f"V must be >= 2, got {V}")
    # matrix rows: rows[1] .. rows[V-1]
    if len(rows) < V:
        raise ValueError(f"expected V-1={V - 1} matrix rows, "
                         f"got {len(rows) - 1} lines after V")
    counts: list[int] = []
    for i in range(V - 1):
        line = rows[1 + i]
        parts = line.split()
        if len(parts) != V - 1 - i:
            raise ValueError(f"matrix row {i} needs {V - 1 - i} integers, "
                             f"got {line!r}")
        for p in parts:
            try:
                m = int(p)
            except ValueError:
                raise ValueError(f"matrix row {i}: non-integer count {p!r}") from None
            if m < 0:
                raise ValueError(f"matrix row {i}: negative count {m}")
            counts.append(m)
    E = sum(counts)
    queue = rows[1 + (V - 1):]
    if len(queue) != E:
        raise ValueError(f"edge queue needs exactly E={E} lines, got {len(queue)}")
    edges: list[tuple[int, int, str]] = []
    slots = _slot_pairs(V)
    t = 0
    for (i, j), m in zip(slots, counts):
        for _ in range(m):
            kind = queue[t]
            t += 1
            if kind not in ("R", "L", "C"):
                raise ValueError(f"edge queue entry must be R|L|C, got {kind!r}")
            edges.append((i, j, kind))
    return edges


def load_topology(path) -> list:
    """Read the topology file from disk."""
    with open(path, "r", encoding="utf-8") as fh:
        return parse_topology(fh.read())

File: AlgorithmLcr/Try3-Component types with known topology and corresponding positions/cppversion/src/iofmt.cpp
#include "iofmt.hpp"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <fstream>
#include <map>
#include <sstream>
#include <stdexcept>

namespace tf {

namespace {

std::vector<std::string> contentLines(const std::string& text) {
    std::vector<std::string> out;
    std::istringstream is(text);
    std::string raw;
    while (std::getline(is, raw)) {
        size_t hash = raw.find('#');
        if (hash != std::string::npos) raw = raw.substr(0, hash);
        size_t a = raw.find_first_not_of(" \t\r");
        size_t b = raw.find_last_not_of(" \t\r");
        if (a == std::string::npos) continue;
        out.push_back(raw.substr(a, b - a + 1));
    }
    return out;
}

[[noreturn]] void fail(const std::string& msg) { throw std::invalid_argument(msg); }

std::string readFile(const std::string& path) {
    std::ifstream fh(path, std::ios::binary);
    if (!fh) fail("cannot open input file: " + path);
    std::ostringstream ss;
    ss << fh.rdbuf();
    return ss.str();
}

std::vector<std::string> splitFields(const std::string& line) {
    std::istringstream ls(line);
    std::vector<std::string> parts;
    std::string tok;
    while (ls >> tok) parts.push_back(tok);
    return parts;
}

std::vector<std::pair<int, int>> slotPairs(int V) {
    std::vector<std::pair<int, int>> out;
    for (int i = 0; i < V; ++i)
        for (int j = i + 1; j < V; ++j) out.push_back({i, j});
    return out;
}

}  // namespace

Measurements parseMeasurements(const std::string& text) {
    auto rows = contentLines(text);
    if (rows.empty()) fail("empty measurement input");
    long n = 0;
    try {
        size_t pos = 0;
        n = std::stol(rows[0], &pos);
        if (pos != rows[0].size()) throw std::invalid_argument("trailing");
    } catch (const std::exception&) {
        fail("first line must be integer n, got '" + rows[0] + "'");
    }
    if (n < 1) fail("n must be a positive integer");
    if ((long)rows.size() - 1 != n)
        fail("n=" + std::to_string(n) + " but " + std::to_string(rows.size() - 1) +
             " data lines follow");
    Measurements out;
    out.f.reserve((size_t)n);
    out.z.reserve((size_t)n);
    for (long k = 0; k < n; ++k) {
        const std::string& line = rows[(size_t)k + 1];
        auto parts = splitFields(line);
        if (parts.size() != 3)
            fail("line " + std::to_string(k + 2) + ": expected 'f Rz Iz', got '" + line + "'");
        double fv = 0, rz = 0, iz = 0;
        try {
            size_t p0, p1, p2;
            fv = std::stod(parts[0], &p0);
            rz = std::stod(parts[1], &p1);
            iz = std::stod(parts[2], &p2);
            if (p0 != parts[0].size() || p1 != parts[1].size() || p2 != parts[2].size())
                throw std::invalid_argument("trailing");
        } catch (const std::exception&) {
            fail("line " + std::to_string(k + 2) + ": non-numeric field in '" + line + "'");
        }
        if (!std::isfinite(fv) || !std::isfinite(rz) || !std::isfinite(iz))
            fail("line " + std::to_string(k + 2) + ": non-finite value");
        if (fv <= 0.0)
            fail("line " + std::to_string(k + 2) + ": frequency must be > 0");
        out.f.push_back(fv);
        out.z.emplace_back(rz, iz);
    }
    return out;
}

Measurements loadMeasurements(const std::string& path) {
    return parseMeasurements(readFile(path));
}

std::string formatMeasurements(const std::vector<double>& f,
                               const std::vector<Complex>& z) {
    if (f.size() != z.size())
        fail("f has " + std::to_string(f.size()) + " points, z has " +
             std::to_string(z.size()));
    if (f.empty()) fail("need at least one measurement point");
    std::ostringstream out;
    char buf[3][32];
    out << f.size() << "\n";
    for (size_t k = 0; k < f.size(); ++k) {
        if (!(std::isfinite(f[k]) && f[k] > 0.0))
            fail("frequency must be positive and finite");
        if (!std::isfinite(z[k].real()) || !std::isfinite(z[k].imag()))
            fail("impedance parts must be finite");
        std::snprintf(buf[0], sizeof(buf[0]), "%.17g", f[k]);
        std::snprintf(buf[1], sizeof(buf[1]), "%.17g", z[k].real());
        std::snprintf(buf[2], sizeof(buf[2]), "%.17g", z[k].imag());
        out << buf[0] << " " << buf[1] << " " << buf[2] << "\n";
    }
    return out.str();
}

std::vector<std::tuple<int, int, char>> parseTopology(const std::string& text) {
    auto rows = contentLines(text);
    if (rows.empty()) fail("empty topology input");
    int V = 0;
    try {
        size_t pos = 0;
        long v = std::stol(rows[0], &pos);
        if (pos != rows[0].size()) throw std::invalid_argument("trailing");
        V = (int)v;
    } catch (const std::exception&) {
        fail("first line must be integer V, got '" + rows[0] + "'");
    }
    if (V < 2) fail("V must be >= 2");
    if ((int)rows.size() < V)
        fail("expected V-1=" + std::to_string(V - 1) + " matrix rows, got " +
             std::to_string((int)rows.size() - 1) + " lines after V");
    std::vector<int> counts;
    for (int i = 0; i < V - 1; ++i) {
        const std::string& line = rows[(size_t)(1 + i)];
        auto parts = splitFields(line);
        if ((int)parts.size() != V - 1 - i)
            fail("matrix row " + std::to_string(i) + " needs " +
                 std::to_string(V - 1 - i) + " integers, got '" + line + "'");
        for (const auto& p : parts) {
            long m = 0;
            try {
                size_t pos = 0;
                m = std::stol(p, &pos);
                if (pos != p.size()) throw std::invalid_argument("trailing");
            } catch (const std::exception&) {
                fail("matrix row " + std::to_string(i) + ": non-integer count '" + p + "'");
            }
            if (m < 0)
                fail("matrix row " + std::to_string(i) + ": negative count " +
                     std::to_string(m));
            counts.push_back((int)m);
        }
    }
    long E = 0;
    for (int m : counts) E += m;
    long queueLen = (long)rows.size() - 1 - (V - 1);
    if (queueLen != E)
        fail("edge queue needs exactly E=" + std::to_string(E) + " lines, got " +
             std::to_string(queueLen));
    std::vector<std::tuple<int, int, char>> edges;
    auto slots = slotPairs(V);
    size_t t = 0;
    for (size_t s = 0; s < slots.size(); ++s) {
        for (int q = 0; q < counts[s]; ++q) {
            const std::string& kind = rows[1 + (size_t)(V - 1) + t];
            ++t;
            if (kind.size() != 1 || (kind[0] != 'R' && kind[0] != 'L' && kind[0] != 'C'))
                fail("edge queue entry must be R|L|C, got '" + kind + "'");
            edges.push_back({slots[s].first, slots[s].second, kind[0]});
        }
    }
    return edges;
}

std::vector<std::tuple<int, int, char>> loadTopology(const std::string& path) {
    return parseTopology(readFile(path));
}

std::string formatTopology(const std::vector<std::tuple<int, int, char>>& edges) {
    if (edges.empty()) fail("need at least one edge");
    int V = 0;
    for (const auto& [u, v, k] : edges) {
        (void)k;
        V = std::max({V, u + 1, v + 1});
    }
    std::vector<std::tuple<int, int, char>> ordered = edges;
    std::stable_sort(ordered.begin(), ordered.end(),
                     [](const auto& a, const auto& b) {
                         int ua = std::min(std::get<0>(a), std::get<1>(a));
                         int va = std::max(std::get<0>(a), std::get<1>(a));
                         int ub = std::min(std::get<0>(b), std::get<1>(b));
                         int vb = std::max(std::get<0>(b), std::get<1>(b));
                         return std::make_pair(ua, va) < std::make_pair(ub, vb);
                     });
    std::map<std::pair<int, int>, int> counts;
    for (const auto& [u, v, k] : ordered) {
        (void)k;
        counts[{std::min(u, v), std::max(u, v)}] += 1;
    }
    std::ostringstream out;
    out << V << "\n";
    for (int i = 0; i < V - 1; ++i) {
        for (int j = i + 1; j < V; ++j) {
            if (j > i + 1) out << " ";
            auto it = counts.find({i, j});
            out << (it == counts.end() ? 0 : it->second);
        }
        out << "\n";
    }
    for (const auto& [u, v, k] : ordered) {
        (void)u;
        (void)v;
        out << k << "\n";
    }
    return out.str();
}

}  // namespace tf
